File: kotlin-node/karakum/plugins/convertReadableSymbol.js
import ts from "typescript";
import * as karakum from "karakum"

export default function (node, context) {
    if (
        ts.isMethodDeclaration(node)
        && ts.isComputedPropertyName(node.name)

        && node.parent
        && ts.isClassDeclaration(node.parent)
        && node.parent.name?.text === "ReadableBase"
    ) {
        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)

        return `/* ${typeScriptService?.printNode(node)} */`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertReadableSymbol.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.isClassDeclaration
import typescript.isComputedPropertyName
import typescript.isMethodDeclaration

val convertReadableSymbol = createPlugin { node, context, _ ->
    nullable {
        ensure(isMethodDeclaration(node))

        val name = node.name
        ensure(isComputedPropertyName(name))

        val classNode = ensureNotNull(node.getParentOrNull())
        ensure(isClassDeclaration(classNode))
        ensure(classNode.name?.text == "Readable")

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

        "/* ${typeScriptService.printNode(node)} */"
    }
}
